### PR TITLE
Cut filter partition based on metadata_block_size

### PR DESCRIPTION
**Pull request by maysamyabandeh**
Currently metadata_block_size controls only index partition size. With this patch a partition is cut after any of index or filter partitions reaches metadata_block_size.

### CODE REVIEW

File: table/block_based_table_builder.cc [hunk 1]
```diff
@@ -76,10 +76,18 @@ FilterBlockBuilder* CreateFilterBlockBuilder(
   } else {
     if (table_opt.partition_filters) {
       assert(p_index_builder != nullptr);
+      // Since after partition cut request from filter builder it takes time
+      // until index builder actully cuts the partition, we take the lower bound
+      // as partition size.
+      uint64_t partition_size =
+          table_opt.metadata_block_size >
+                  (uint64_t)table_opt.block_size_deviation
+              ? table_opt.metadata_block_size - table_opt.block_size_deviation
```
Comment (yiwu-arbug) [inline table/block_based_table_builder.cc]: metadata_block_size * (1 - block_size_deviation)?

File: table/index_builder.cc [hunk 2]
```diff
@@ -101,7 +106,7 @@ void PartitionedIndexBuilder::AddIndexEntry(
     if (sub_index_builder_ != nullptr) {
       std::string handle_encoding;
       block_handle.EncodeTo(&handle_encoding);
-      bool do_flush =
+      bool do_flush = partition_cut_requested_ ||
```
Comment (yiwu-arbug) [inline table/index_builder.cc]: do we need to reset partition_cut_requested_ to false?

File: include/rocksdb/filter_policy.h [hunk 3]
```diff
@@ -41,6 +42,64 @@ class FilterBitsBuilder {
...
   // The ownership of actual data is set to buf
   virtual Slice Finish(std::unique_ptr<const char[]>* buf) = 0;
+
+  // Calculate num of entries fit into a space.
+  virtual int CalculateNumEntry(const uint32_t space) {
+    throw new std::runtime_error("CalculateNumEntry not Implemented");
+  }
+};
+
+class FullFilterBitsBuilder : public FilterBitsBuilder {
```
Comment (maysamyabandeh) [inline include/rocksdb/filter_policy.h]: That is a big decision. So far the class was totally hidden. I moved here so I could at least test it. If we moved it to internal header then we would have an obligation to maintain it forever, for which there does not seem to be a strong motivation.

File: table/block_based_table_builder.cc (hunk 1 above)
Comment (maysamyabandeh) [inline table/block_based_table_builder.cc]: block_size_deviation is not a fraction, it is rather in number of bytes +/- the target size.

File: table/index_builder.cc (hunk 2 above)
Comment (maysamyabandeh) [inline table/index_builder.cc]: It is reset each time MakeNewSubIndexBuilder  is called, which is also called below after the flush.

File: include/rocksdb/filter_policy.h (hunk 3 above)
Comment (yiwu-arbug) [inline include/rocksdb/filter_policy.h]: This is external (user facing) header. I'm suggesting to move this to internal (hidden) header.

File: table/block_based_table_builder.cc (hunk 1 above)
Comment (yiwu-arbug) [inline table/block_based_table_builder.cc]: The comment in table.h suggest the opposite: https://github.com/facebook/rocksdb/blob/master/include/rocksdb/table.h#L131-L136
Can you check?

File: table/index_builder.cc (hunk 2 above)
Comment (yiwu-arbug) [inline table/index_builder.cc]: Got it.

File: table/block_based_table_builder.cc (hunk 1 above)
Comment (maysamyabandeh) [inline table/block_based_table_builder.cc]: Ah you are right. Thanks.

File: include/rocksdb/filter_policy.h (hunk 3 above)
Comment (maysamyabandeh) [inline include/rocksdb/filter_policy.h]: :) yeah i got it backward.

**Comment by yiwu-arbug** [comment]
Both AppVoyor and Travis are failing.

**Pull request closed without merging**